noodles-bed/src/de.rs

```diff
@@ -1,934 +1,110 @@
-                    start = Some(map.next_value()?);
-                }
-                Field::End => {
-                    if end.is_some() {
-                        return Err(de::Error::duplicate_field("end"));
-                    }
-                    end = Some(map.next_value()?);
-                }
-            }
-        }
-        let chrom: &str = chrom.ok_or_else(|| de::Error::missing_field("chrom"))?;
-        let start: usize = start.ok_or_else(|| de::Error::missing_field("start"))?;
-        let end: usize = end.ok_or_else(|| de::Error::missing_field("end"))?;
-
-        Record::<3>::builder()
-            .set_reference_sequence_name(chrom)
-            .set_start_position(Position::try_from(start).map_err(|err| de::Error::custom(err))?)
-            .set_end_position(Position::try_from(end).map_err(|err| de::Error::custom(err))?)
-            .build().map_err(|err| de::Error::custom(err))
-    }
-}
-
-
-impl<'de> Deserialize<'de> for Record<3> {
-    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        deserializer.deserialize_struct("Record", FIELDS, Record3Visitor)
-    }
-}
-
-pub fn from_reader<R>(reader: &mut Reader<R>) -> std::result::Result<Vec<Record<3>>, std::io::Error>
+use crate::de::RecordState::ExpectingChrom;
+use crate::{error, Reader, Record};
+use error::{Error, Result};
+use serde::de::{DeserializeSeed, SeqAccess, Visitor};
+use serde::{de, forward_to_deserialize_any, Deserialize, Deserializer};
+use std::io;
+use std::iter::Peekable;
+
+/// A user would likely use a function such as the following to deserialize a set of records.
+pub fn from_bytes<'a, T>(records: &'a [u8]) -> Result<T>
 where
-    R: BufRead
+    T: Deserialize<'a>,
 {
-    reader
-        .records::<3>()
-        .collect()
-}
-
-pub struct Deserializer<'de> {
-    // This string starts with the input data and characters are truncated off
-    // the beginning as data is parsed.
-    input: &'de str,
+    let mut reader = Reader::new(records);
+    let records = reader.records();
+    let mut deserializer = Record3Deserializer::new(records);
+    T::deserialize(&mut deserializer)
 }
 
-impl<'de> Deserializer<'de> {
-    // By convention, `Deserializer` constructors are named like `from_xyz`.
-    // That way basic use cases are satisfied by something like
-    // `serde_json::from_str(...)` while advanced use cases that require a
-    // deserializer can make one with `serde_json::Deserializer::from_str(...)`.
-    pub fn from_str(input: &'de str) -> Self {
-        Deserializer { input }
-    }
+/// This leeps track of the state associated with the next record that is parsed.
+/// We need to keep track of the state as records are returned whole from the iterator, whereas
+/// the deserializer progresses one field at a time.
+enum RecordState {
+    ExpectingChrom,
+    ExpectingChromStart(Record<3>),
+    ExpectingChromEnd(Record<3>),
 }
```

Comment:
Keeping track of the `Deserializer` state is kind of like the inverse of keeping track of the [`bed::record::builder::Builder`](https://docs.rs/noodles-bed/0.4.0/noodles_bed/record/builder/struct.Builder.html) state in the `Serializer`.